Android/Samples/libraries/kml/src/main/java/com/microsoft/maps/kml/KMLParser.java

```diff
@@ -195,10 +199,79 @@ private MapPolyline parseLineString()
         skipToEndOfTag();
       }
     }
-    verifyCoordinatesParsed(hasParsedCoordinates);
+    verifyElementSeen("coordinates", hasParsedCoordinates);
     return line;
   }
 
+  @NonNull
+  private MapPolygon parsePolygon() throws IOException, XmlPullParserException, KMLParseException {
+    mParser.require(XmlPullParser.START_TAG, mNameSpace, "Polygon");
+    MapPolygon polygon = mFactory.createMapPolygon();
+    ArrayList<ArrayList<Geoposition>> rings = new ArrayList<>();
+    boolean hasOuterBoundary = false;
+    AltitudeReferenceSystemWrapper altitudeReferenceSystemWrapper =
+        new AltitudeReferenceSystemWrapper(AltitudeReferenceSystem.GEOID);
+    while (mParser.next() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("outerBoundaryIs") || type.equals("innerBoundaryIs")) {
+        if (type.equals("outerBoundaryIs")) {
+          verifyElementNotSeen("outerBoundaryIs", hasOuterBoundary);
+          hasOuterBoundary = true;
+        }
+        ArrayList<Geoposition> positions = parsePolygonRing(type, altitudeReferenceSystemWrapper);
+        rings.add(positions);
+      } else {
+        skipToEndOfTag();
+      }
+    }
+    verifyElementSeen("outerBoundaryIs", hasOuterBoundary);
+    ArrayList<Geopath> paths = new ArrayList<>(rings.size());
+    for (ArrayList<Geoposition> ring : rings) {
+      ParsingHelpers.setAltitudesToZeroIfAtSurface(
+          ring, altitudeReferenceSystemWrapper.getAltitudeReferenceSystem());
+      paths.add(new Geopath(ring, altitudeReferenceSystemWrapper.getAltitudeReferenceSystem()));
+    }
+    polygon.setPaths(paths);
+    return polygon;
+  }
+
+  /* ArrayList positions is initialized by parseCoordinates, or an error is thrown if no
+   * <coordinates> tag is present. */
+  @NonNull
+  private ArrayList<Geoposition> parsePolygonRing(
+      @NonNull String tag, @NonNull AltitudeReferenceSystemWrapper altitudeReferenceSystemWrapper)
+      throws IOException, XmlPullParserException, KMLParseException {
+    mParser.require(XmlPullParser.START_TAG, mNameSpace, tag);
+    mParser.nextTag();
+    mParser.require(XmlPullParser.START_TAG, mNameSpace, "LinearRing");
+    ArrayList<Geoposition> positions = null;
+    boolean hasParsedCoordinates = false;
+    while (mParser.next() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("coordinates")) {
+        verifyElementNotSeen(type, hasParsedCoordinates);
+        positions = parseCoordinates(altitudeReferenceSystemWrapper);
+        String exceptionMessage = ParsingHelpers.verifyPolygonRing(positions);
```

Comment:
The naming of verifyPolygonRing is confusing. All the other verify methods throw an exception if it's not valid and this one doesn't; it instead returns a string that if not null is the error message if the input is invalid.